Crunch03 (#131)
alem-platform/bootcamp-go: crunch03/README.md

The program should henadle following flags:

- Flag `--help` - provides information about the program's capabilities and shows all implemented flags.
- Flag `--verbose` - displays information about the map size, the number of ticks, speed, map name, etc. Information is shown at the top.
- Flag `--delay-ms=` - sets the animation speed in milliseconds. The default speed is `2500` milliseconds.

### Input Format

The input consists of a grid of characters representing the initial state of the game, followed by optional flags.

### Game Rules

1. **Grid Evolution:**

   - Any live cell with fewer than two live neighbors dies (underpopulation).
   - Any live cell with two or three live neighbors lives on to the next generation.
   - Any live cell with more than three live neighbors dies (overpopulation).
   - Any dead cell with exactly three live neighbors becomes a live cell (reproduction).
   - Overall, the grid evolution shoild satisfy the rules of `Game of Life`.

2. **Edge Cases:**

   - Handle minimum size grids (3x3) and larger grids.
   - Handle improper inputs with error message `Error: invalid input.`.
   - Terminate the program if the input is invalid.

>    - If options conflict, then accept only the first option.

Author: ```suggestion
   - If the flags conflict, of the conflicting flags, accept the first one.
```

>    - If the flags conflict, of the conflicting flags, accept the first one.

Author: ```suggestion
   - If the flags conflict, of the conflicting flags, accept the first one.
   - The program should run until there are no living cells left on the grid.
```

```sh
Tick: 3
Grid Size: 10x10
Live Cells: 29
DelayMs: 1000ms

· · · · · · · · · ·
× · · × · × · · · ·
· × · · · × · · · ·
× · × · · × · · · ·
× · · · · × · · · ·
× · · · · × · · · ·
× · · · · × · · · ·
· × · · · × · · · ·
· · × · × · · · · ·
· · · × · · · · · ·
```

**Output After Three Ticks:**

```sh
Tick: 4
Grid Size: 10x10
Live Cells: 25
DelayMs: 1000ms

· · · × · · · · · ·
× · · · × · × · · ·
· × · · · × · · · ·
× · × · · · · · · ·
× · · · · × · · · ·
× · · · · × · · · ·
× · · · · × · · · ·
· × · · · × · · · ·
· · × · · · · · · ·
· · · · × · · · · ·
```

### Bonus Features

- **Flag `--file=`**: Reads the initial grid from a specified file.

  - **Example Usage:**
    ```sh
    student$: go run main.go --file=path/to/grid.txt
    ```

- **Flag `--edges-portal`**: Enables portal edges where cells that exit the grid appear on the opposite side.

  - **Example Usage:**
    ```sh
    student$: go run main.go --edges-portal
    ```

- **Flag `--random=WxH`**: Generates a random grid of the specified width (W) and height (H).

  - **Example Usage:**
    ```sh
    student$: go run main.go --random=5x5
    · · × · ·
    × · × · ·
    · × × · ·
    × · · · ·
    · · · × ×
    ```

> - **Flag `--fullscreen`**: Adjusts the grid to fit the terminal size with empty cells. Ff the terminal size is smaller, it may cut off the entered card. The program should take into account the size of the terminal only at the time of launch.

Author: Есть ошибка

# crunch03

| Available                                     |
| --------------------------------------------- |
| `*` (except packages that implement the game) |

**Task**

Implement the `Game-of-Life` simulation.

Read more about the game here[Game of Life](https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life)

See an example of the game principles here: [Game of Life Example](https://conwaylife.com/)

The program should accept the following symbols: `#.` where `#` represents a live cell and `.` represents an empty cell.

- Use `×` to display live cells;
- Use `·` to display dead cells;

> The program should henadle following flags:

Author: ```suggestion
The program should handle following flags:
```

- Flag `--help` - provides information about the program's capabilities and shows all implemented flags.
- Flag `--verbose` - displays information about the map size, the number of ticks, speed, map name, etc. Information is shown at the top.
- Flag `--delay-ms=` - sets the animation speed in milliseconds. The default speed is `2500` milliseconds.

### Input Format

The input consists of a grid of characters representing the initial state of the game, followed by optional flags.

### Game Rules

1. **Grid Evolution:**

   - Any live cell with fewer than two live neighbors dies (underpopulation).
   - Any live cell with two or three live neighbors lives on to the next generation.
   - Any live cell with more than three live neighbors dies (overpopulation).
   - Any dead cell with exactly three live neighbors becomes a live cell (reproduction).
   - Overall, the grid evolution shoild satisfy the rules of `Game of Life`.

2. **Edge Cases:**

   - Handle minimum size grids (3x3) and larger grids.

>    - Handle improper inputs with error message `Error: invalid input.`.

Author: ```suggestion
```

The program should accept the following symbols: `#.` where `#` represents a live cell and `.` represents an empty cell.

- Use `×` to display live cells;
- Use `·` to display dead cells;

The program should henadle following flags:

- Flag `--help` - provides information about the program's capabilities and shows all implemented flags.
- Flag `--verbose` - displays information about the map size, the number of ticks, speed, map name, etc. Information is shown at the top.
- Flag `--delay-ms=` - sets the animation speed in milliseconds. The default speed is `2500` milliseconds.

### Input Format

The input consists of a grid of characters representing the initial state of the game, followed by optional flags.

### Game Rules

1. **Grid Evolution:**

   - Any live cell with fewer than two live neighbors dies (underpopulation).
   - Any live cell with two or three live neighbors lives on to the next generation.
   - Any live cell with more than three live neighbors dies (overpopulation).
   - Any dead cell with exactly three live neighbors becomes a live cell (reproduction).
   - Overall, the grid evolution shoild satisfy the rules of `Game of Life`.

2. **Edge Cases:**

> 

Author: В случае если опции конфликтуют, то принимать только первую опцию

   - Handle minimum size grids (3x3) and larger grids.
   - Handle improper inputs with appropriate error messages and handling.
   - Terminate the program if the input is invalid.

> 

Author: Единственное требования к ошибка, это они должны быть понятными, и описывать то что не верно.

Reviewer: - Handle improper inputs with error message `Error: invalid input.`.

3. **Verbose Mode:**
   - In verbose mode, display the following information at the top of each tick:
     - Tick number.
     - Grid size.
     - Number of live cells.
     - Delay.

### Example 1: Basic Input and Evolution

```sh
student$: go run main.go
```

**Grid Input:**

```sh
4 4
....
.##.
.##.
....

> ```

Author: Обьеденить

**Initial Output:**

```sh
· · · ·
· × × ·
· × × ·
· · · ·
```

**Output After One Tick:**

```sh
· · · ·
· × × ·
· × × ·
· · · ·
```

**Example 2: Edge Case with Minimum Grid Size and Invalid Input**

**Input:**

```sh
go run main.go

> ```

Author: Нет текста ошибки

**Grid Input:**

```
2 2
..
..

> ```

Author: Будто бы нужно обьеденить

**Output:**

```sh
Error: invalid input.
```

### Example 3: Using the `--help` Flag

> 

Author: Дописать что это пример вывода, вывод может измененным, главное чтобы он соответствовал требованию `help`

**Input:**

```sh
go run main.go --help
```

**Output:**

```sh
Usage: go run main.go [options]

Options:
  --help        : Show the help message and exit
  --verbose     : Display detailed information about the simulation, including grid size, number of ticks, speed, and map name
  --delay-ms=<n>: Set the animation speed in milliseconds. Default is 2500 milliseconds

> ```

Author: Обьеденить

### Example 4: Input with Verbose Mode

> 

Author: Тоже самое для verbose

**Input:**

```sh
go run main.go --verbose
```

**Grid Input:**

```
6 6
......
..##..
.##...
..##..
..##..
......

> ```

Author: Обьеденить

**Initial Output:**

```sh
Tick: 1
Grid Size: 6x6
Live Cells: 8
Delay: 2500ms

· · · · · ·
· · × × · ·
· × × · · ·
· · × × · ·
· · × × · ·
· · · · · ·
```

**Output After One Tick:**

```sh
Tick: 2
Grid Size: 6x6
Live Cells: 9
Delay: 2500ms

· · · · · ·
· × × × · ·
× · · × · ·
· · · × · ·
· × × · · ·
· · · · · ·
```

**Output After Two Ticks:**

```sh
Tick: 3
Grid Size: 6x6
Live Cells: 11
Delay: 2500ms

· × · · · ·
× · · · × ·
· · × · · ×
× · · · · ×
· × · × × ·
· · · · · ·
```

### Example 5: Input with Verbose Mode and Custom Animation Delay

**Input:**

```sh
go run main.go --verbose --delay-ms=1000
```

**Grid Input:**

```

> 10 10

Author: Обьеденить

· · · × × · · · · ·
· · · · · · · · · ·
```

**Output After Two Ticks:**

```sh
Tick: 3
Grid Size: 10x10
Live Cells: 29
Delay: 1000ms

· · · · · · · · · ·
× · · × · × · · · ·
· × · · · × · · · ·
× · × · · × · · · ·
× · · · · × · · · ·
× · · · · × · · · ·
× · · · · × · · · ·
· × · · · × · · · ·
· · × · × · · · · ·
· · · × · · · · · ·
```

**Output After Three Ticks:**

```sh
Tick: 4
Grid Size: 10x10
Live Cells: 25
Delay: 1000ms

· · · × · · · · · ·
× · · · × · × · · ·
· × · · · × · · · ·
× · × · · · · · · ·
× · · · · × · · · ·
× · · · · × · · · ·
× · · · · × · · · ·
· × · · · × · · · ·
· · × · · · · · · ·
· · · · × · · · · ·
```

### Bonus Features

- **Flag `--file=`**: Reads the initial grid from a specified file.

  - **Example Usage:**
    ```sh
    go run main.go --file=path/to/grid.txt
    ```

- **Flag `--edges-portal`**: Enables portal edges where cells that exit the grid appear on the opposite side.

  - **Example Usage:**
    ```sh
    go run main.go --edges-portal
    ```

- **Flag `--random=WxH`**: Generates a random grid of the specified width (W) and height (H).

  - **Example Usage:**
    ```sh
    go run main.go --random=5x5
    ```
  - **Initial Output:**

>     ```

Author: Обьеденить

    · · × · ·
    × · × · ·
    · × × · ·
    × · · · ·
    · · · × ×
    ```

> - **Flag `--fullscreen`**: Adjusts the grid to fit the terminal size with empty cells.

Author: Стоит уточнить что в случае если размер терминала будет меньше, то он может обрезать введенную карту. Программа должна учитывать размер терминала только в момент запуска

- Flag `--verbose` - displays information about the map size, the number of ticks, speed, map name, etc. Information is shown at the top.
- Flag `--delay-ms=` - sets the animation speed in milliseconds. The default speed is `2500` milliseconds.

### Input Format

The input consists of a grid of characters representing the initial state of the game, followed by optional flags.

### Game Rules

1. **Grid Evolution:**

   - Any live cell with fewer than two live neighbors dies (underpopulation).
   - Any live cell with two or three live neighbors lives on to the next generation.
   - Any live cell with more than three live neighbors dies (overpopulation).
   - Any dead cell with exactly three live neighbors becomes a live cell (reproduction).
   - Overall, the grid evolution shoild satisfy the rules of `Game of Life`.

2. **Edge Cases:**

   - Handle minimum size grids (3x3) and larger grids.
   - Handle improper inputs with appropriate error messages and handling.
   - Terminate the program if the input is invalid.

3. **Verbose Mode:**
   - In verbose mode, display the following information at the top of each tick:
     - Tick number.
     - Grid size.
     - Number of live cells.

>      - Current speed.

Author: Переименовать, так как скорость общее понятие. DelayMs
